File: CMSC137/Code/LabExer6.py
# 10/22/23
# LabExer6

# >>>>>>>>>>>>>>>>>>>>>>>>>> helper functions <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
def checksum_binary_add(bin1, bin2, max_len):
    carry = ""
    str_bin = str(bin(int(bin1,2) + int(bin2,2)))[2:]
    if len(str_bin) > max_len:
        carry = str_bin[0]
        str_bin = str_bin[1:]
    if carry != "":
        str_bin = bin(int(str_bin, 2) + int(carry,2))[2:].zfill(max_len)

    return str(str_bin)

# >>>>>>>>>>>>>>>>>>>>>>>>>> user functions <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
def parity_check():
    dataword: str = input()
    codeword: str = input()

    dword_count: int = dataword.count("1")

    gen_codeword = str(dataword)
    if dword_count % 2 == 0:
        gen_codeword += "0"
    elif dword_count % 2 != 0:
        gen_codeword += "1"
        
    print(f"Codeword:{gen_codeword}")
    if codeword == gen_codeword:
        print("Data word: Accepted")
    else:
        print("Data word: Discarded")

def twod_parity_check():
    input_str = input()
    bytewords = [input_str[i:i+9] for i in range(0, len(input_str), 9)]
    col_count = [0] * len(bytewords[0])
    row_count = []
    error_count = 0

    for i in range(0,len(bytewords) - 1):
        rc = 0
        for j in range(0,8):
            if bytewords[i][j] == "1":
                col_count[j] += 1
                rc += 1
        row_count.append(rc)

    lc_bits = ""
    lr_bits = ""
    for i in range(0,8):
        curr = bytewords[4][i]
        lr_bits += curr
        if (curr == "1" and col_count[i] % 2 == 0) or (curr == "0" and col_count[i] % 2 != 0):
            error_count += 1

    for i in range(0,4):
        curr = bytewords[i][8]
        lc_bits += curr
        if (curr == "1" and row_count[i] % 2 == 0) or (curr == "0" and row_count[i] % 2 != 0):
            error_count += 1

    if (lc_bits.count("1") % 2 == 0 and bytewords[4][8] == "1") or (lc_bits.count("1") % 2 != 0 and bytewords[4][8] == "0"):
        error_count += 1

    if (lr_bits.count("1") % 2 == 0 and bytewords[4][8] == "1") or (lr_bits.count("1") % 2 != 0 and bytewords[4][8] == "0"):
        error_count += 1
    print(f"Error count: {error_count}")


def checksum():
    input_str: str = input()

    bytewords = [input_str[i:i+7] for i in range(0, len(input_str), 7)]
    sum = bytewords[0] 
    for b in bytewords[1:]:
        sum = checksum_binary_add(sum, b, 7)

    print(sum)
    # if "0" not in sum:
    #     print("Accept data")
    # else:
    #     print("Reject data")

def crc():
    data: str = input()
    keyword: str = input()

    rem = bin(int(data,2) // int(keyword,2))
    if keyword == str(bin(int(data,2) // int(rem,2)))[2:].zfill(len(keyword)):
        print("Accept data")
    else:
        print("CRC error detected")


# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> main <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
def main():
    input_num: str = input()
    if input_num == "1":
        parity_check()
    elif input_num == "2":
        twod_parity_check()
    elif input_num == "3":
        checksum()
    elif input_num == "4":
        crc()

if __name__ == "__main__":
    main()

File: CMSC137/Code/LabExer6.rs
use std::io;

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_line(&mut input)
        .expect("Failed to read from stdin");

    let clean_input = input.trim().to_lowercase();

    match clean_input.as_str() {
        "1" => simple_parity_check(),
        "2" => twod_parity_check(),
        "3" => checksum(),
        "4" => crc(),
        _ => {
            println!("Invalid input. Please enter valid integer (1-4)");
        }
    }
}

// ----------------  helper functions -------------------------

fn validate_bits(input: &str) -> i32 {
    let mut ones_count = 0;

    for c in input.chars() {
        match c {
            '0' => continue,
            '1' => ones_count += 1,
            _ => panic!("Illegal character: '{}'", c),
        }
    }
    ones_count
}

// fn calculate_errors(matrix: &Vec<Vec<char>>) -> i32 {};

// ----------------  user functions -------------------------

fn simple_parity_check() {
    let mut dataword = String::new();
    let mut codeword = String::new();

    io::stdin()
        .read_line(&mut dataword)
        .expect("Failed to read dataword");

    io::stdin()
        .read_line(&mut codeword)
        .expect("Failed to read codeword");

    let dword_count = validate_bits(dataword.trim());
    let cword_count = validate_bits(codeword.trim());
    dataword = dataword.trim().to_string();
    codeword = codeword.trim().to_string();

    if dword_count % 2 == 0 {
        dataword.push_str("0");
        println!("Codeword: {}", dataword);
    } else if dword_count % 2 != 0 {
        dataword.push_str("1");
        println!("Codeword: {}", dataword);
    }

    if cword_count % 2 == 0 {
        codeword.pop().unwrap();
        println!("Data word: {}", codeword);
    } else {
        println!("Data word: Discarded");
    }
}

fn twod_parity_check() {
    let mut input_string = String::new();
    io::stdin()
        .read_line(&mut input_string)
        .expect("Failed to read input string");
    input_string = input_string.trim().to_string();

    if input_string.len() < 45 {
        panic!(
            "Input string must have 45 characters: len={}",
            input_string.len()
        );
    }

    let temp_vec: Vec<char> = input_string.chars().collect();
    let mut matrix: Vec<Vec<char>> = vec![vec![' '; 9]; 5];
    let mut col_count: Vec<i32> = vec![0; 8];
    let mut row_count: Vec<i32> = vec![0; 4];

    let mut row = 0;
    let mut col = 0;
    // let mut cc = 0;
    let mut rc = 0;
    for ch in temp_vec.iter() {
        matrix[row][col] = *ch;
        if *ch == '1' && col < 8 && row < 4 {
            rc += 1;
            col_count[col] += 1;
        }
        col += 1;
        if col == 9 {
            col = 0;
            if row < 4 {
                row_count[row] = rc;
            }
            rc = 0;
            row += 1;
        }
        if row == 5 {
            break;
        }
    }
    println!("row_count = {:?}\ncol_count = {:?}", row_count, col_count);

    for row in &matrix {
        for char in row {
            print!("{} ", char);
        }
        println!();
    }

    let mut error_count = 0;

    for i in 0..8 {
        if (matrix[4][i] == '1' && col_count[i] % 2 == 0)
            || (matrix[4][i] == '0' && col_count[i] % 2 != 0)
        {
            error_count += 1;
        }
    }
    for i in 0..4 {
        if (matrix[i][8] == '1' && row_count[i] % 2 == 0)
            || (matrix[i][8] == '0' && row_count[i] % 2 != 0)
        {
            error_count += 1;
        }
    }
    let mut final_rc = 0;
    let mut final_cc = 0;
    for i in 0..8 {
        if matrix[4][i] == '1' {
            final_rc += 1;
        }
    }
    for i in 0..4 {
        if matrix[i][8] == '1' {
            final_cc += 1;
        }
    }

    if (final_rc % 2 == 0 && matrix[4][8] == '1') || (final_rc % 2 != 0 && matrix[4][8] == '0') {
        error_count += 1;
    }
    if (final_cc % 2 == 0 && matrix[4][8] == '1') || (final_cc % 2 != 0 && matrix[4][8] == '0') {
        error_count += 1;
    }
    println!("Error count: {}", error_count);
}

fn checksum() {
    let mut rcv_msg = String::new();
    io::stdin()
        .read_line(&mut rcv_msg)
        .expect("Failed to read input string");

    let k = 8;
    let c1 = &rcv_msg[0..k];
    let c2 = &rcv_msg[k..2 * k];
    let c3 = &rcv_msg[2 * k..3 * k];
    let c4 = &rcv_msg[3 * k..4 * k];
    let checksum = &rcv_msg[4 * k..5 * k];

    println!("cheksum = {}", checksum);
}

fn crc() {
    println!("crc");
}
